services/execsrv/events.go

```diff
@@ -38,13 +39,24 @@ type StateObserver interface {
 	// Name returns the name of the observer and should identify the function of the observer.
 	Name() string
 	// Update handles the state change of the exec.
-	Update(state State)
+	// The update could have changed the state, the
+	// new state should be returned to the informer
+	// which will dispatch the new state.
+	Update(state State) State
```

Comment:
This changes the behavior of the informer-observer pattern right? So this relies on the observer to figure out if the state has changed?

Reply:
Not quite (if I understood the current pattern well enough). 

Before we used to have a few things
1. An `Observer` was built, and held a reference to it's `Informer` 
    ```go
    type StateObserverFunc func(exec types.Exec, inf StateInformer) StateObserver
    ```
2. An `Informer`'s `Inform` function carried an `id` -- which is the id of the observer that should be informed. This was regularly confused with the exec id. This ID (which is the ID of the observer that the informer should inform) leaks details about which observers are registered with an informer. AND was confusing because generally _all_ observers should be updated for an informer event, not just a specific one.
    ```go
    type StateInformer interface {
    	Inform(id string, state State)
    	// trimmed
    }
    ```
3. A few of the observers would change the state of an Exec (maybe observe it was initialising, do health checks, and promote it to running). This is why in number 1 the observer held a reference to it's Informer. The observer would call:
    ```go
    func (o Observer) Update(state State) {
        // do some things
        o.informer.Inform(state.ExecID, newState) // this is wrong
    }
    ```
    
In 3, this is wrong for a few reasons: 
- The Inform function was being called with an ExecID, and not an observer ID. So what we wanted to inform wasn't happening.
- The Inform function (even if it received the correct observer ID) would be specific to a single observer, and wouldn't be propagated to all the other observers -- because it would only operate on the single observer ID passed in the inform function.

So given that some of the observers are changing the state of the exec, I figured the best thing would be to return the current/new state, and for the informer to check if that state has changed. If the state has changed, the Informer can be responsible for propagating that state change across _all_ it's registered observers. This fixes two bugs: 
1. That only a single observer would have seen the state changes
2. That we were trying to call observers with an exec ID.